lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -0,0 +1,148 @@
+ * limitations under the License.
+ */
+
+#include "glow/Runtime/HostManager/HostManager.h"
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Graph/Context.h"
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Runtime/Provisioner/Provisioner.h"
+#include "glow/Runtime/RuntimeTypes.h"
+// #include "glow/Runtime/Executor.h"
+
+#include <future>
+#include <queue>
+
+using namespace glow;
+using namespace runtime;
+
+HostManager::HostManager(const std::vector<DeviceConfig> &configs) {
+  DeviceIDTy deviceCount = 0;
+  for (auto &config : configs) {
+    auto newDevice = DeviceManager::createDeviceManager(config.backendKind,
+                                                        config.deviceName);
+    devices_.emplace(deviceCount, newDevice);
+    deviceCount++;
+  }
+  // executor_ = Executor(devices_);
+}
+
+HostManager::~HostManager() { clearHost(); }
+
+ResultCode HostManager::addNetwork(Module *M) {
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  std::vector<DeviceInfo> deviceInfo;
+  for (auto &device : devices_) {
+    DeviceInfo info = DeviceInfo();
+    info.availableMemory = device.second->getAvailableMemory();
+    deviceInfo.push_back(info);
+  }
+  auto partitioner = Partitioner(M, deviceInfo);
+  auto provisioner = llvm::make_unique<Provisioner>();
+  auto nodeList = std::move(partitioner.Partition());
+  auto result = provisioner->provision(nodeList.roots, devices_, *M);
+
+  for (auto &node : nodeList.roots) {
+    roots_.emplace(node->name, std::move(node));
+  }
+  for (auto &node : nodeList.nodes) {
+    networks_.emplace(node->name, std::move(node));
+  }
+  return result;
+}
+
+void HostManager::removeNetwork(llvm::StringRef networkName) {
+  // Walk the tree for the given function, calling evict function for each node
+  // before removing it from networks_ which frees the node.
+  std::lock_guard<std::mutex> networkLock(networkLock_);
+  if (roots_.find(networkName) == roots_.end()) {
+    return;
+  }
+  std::queue<std::string> nodes;
+  std::set<std::string> allNodes;
+  for (auto &child : roots_[networkName]->children) {
+    nodes.push(child->name);
+    allNodes.insert(child->name);
+  }
+
+  while (!nodes.empty()) {
+    auto nodeName = nodes.front();
+    nodes.pop();
+    auto it = networks_.find(nodeName);
+    if (it != networks_.end()) {
+      for (auto &node : it->second->children) {
+        auto name = node->name;
+        if (allNodes.find(node->name) == allNodes.end()) {
+          nodes.push(node->name);
+          allNodes.insert(node->name);
+        }
+      }
+    }
+  }
+  roots_.erase(roots_.find(networkName));
+  for (auto &networkName : allNodes) {
+    auto networkIterator = networks_.find(networkName);
+    if (networkIterator != networks_.end()) {
+      devices_[networkIterator->second->deviceID]->evictNetwork(networkName);
+      networks_.erase(networkIterator);
+    }
+  }
+}
+
+bool HostManager::networkAdded(llvm::StringRef networkName) {
```

Comment:
Can I ask where is this function called?

Reply:
The onnxifi interface has a similar function. The thought is it would be called from there.